### PR TITLE
Add more workflows

### PR DESCRIPTION
**Pull request by ljleb**
Add more workflows:
- preprocess_img2img
- preprocess_latent_img2img
- postprocess_latent_img2img
- postprocess_latent_txt2img

Start designing the api for other extensions to use.

### CODE REVIEW

File: lib_comfyui/external_code.py [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+from typing import List, Tuple, Union, Optional
+from lib_comfyui import global_state, ipc
+
+
+ALL_TABS = ('txt2img', 'img2img')
+Tabs = Union[str, Tuple[str]]
+
+
+@dataclasses.dataclass
+class Workflow:
```
Comment (ljleb) [inline lib_comfyui/external_code.py:13]: note to self: rename to `WorkflowType` or something like that.

File: scripts/comfyui.py [hunk 2]
```diff
@@ -1,52 +1,124 @@
...
-from lib_comfyui import webui_callbacks, webui_settings, global_state, platform_utils
-from comfyui_custom_nodes import webui_postprocess_input, webui_postprocess_output
+from lib_comfyui import webui_callbacks, webui_settings, global_state, platform_utils, external_code
 from lib_comfyui.polling_client import ComfyuiNodeWidgetRequests
-from modules import shared, ui
+from modules import shared, ui, sd_samplers
+
+
+def add_default_workflows():
+    workflows_dir = Path(scripts.basedir(), 'workflows', 'default')
```
Comment (ljleb) [inline scripts/comfyui.py:14]: default_workflows_dir =

File: scripts/comfyui.py [hunk 3]
```diff
@@ -1,52 +1,124 @@
...
 import modules.scripts as scripts
-from lib_comfyui import webui_callbacks, webui_settings, global_state, platform_utils
-from comfyui_custom_nodes import webui_postprocess_input, webui_postprocess_output
+from lib_comfyui import webui_callbacks, webui_settings, global_state, platform_utils, external_code
 from lib_comfyui.polling_client import ComfyuiNodeWidgetRequests
-from modules import shared, ui
+from modules import shared, ui, sd_samplers
+
+
+def add_default_workflows():
```
Comment (ljleb) [inline scripts/comfyui.py:13]: move this to a different file. in fact, split this entire file into smaller files inside lib_comfyui

File: javascript/compute-height.js [hunk 4]
```diff
@@ -4,23 +4,32 @@ function uuidv4() {
...
+
+function changeCurrentWorkflow(workflowTypes, newWorkflowName) {
+    const newWorkflowType = JSON.parse(workflowTypes)[newWorkflowName];
+    const newIFrameElement = document.querySelector(`#comfyui_${newWorkflowType}`);
+    const oldIFrameElement = newIFrameElement.parentElement.querySelector(".comfyui-embedded-widget-display");
+    oldIFrameElement.classList.remove("comfyui-embedded-widget-display");
+    newIFrameElement.classList.add("comfyui-embedded-widget-display");
+}
+
+const WORKFLOW_IDS = [
```
Comment (ljleb) [inline javascript/compute-height.js:17]: Read this list from a gradio element, but only once the element is defined.

File: scripts/comfyui.py [hunk 5]
```diff
@@ -1,52 +1,124 @@
...
+    workflows_dir = Path(scripts.basedir(), 'workflows', 'default')
+
+    workflows = [
+        external_code.Workflow(
+            base_id='sandbox_tab',
+            display_name='Sandbox',
+            tabs=(),
```
Comment (ljleb) [inline scripts/comfyui.py:20]: this workflow is pretty awkward to define. might need redesign?

Comment (ljleb) [reply scripts/comfyui.py:20]: maybe should rename "sandbox" to "standalone"?

File: lib_comfyui/polling_client.py [hunk 6]
```diff
@@ -133,6 +135,16 @@ async def set_polling_server_focused_webui_client_id(request):
         print(f'[sd-webui-comfyui] set client webui_client_id - \n{webui_client_id}')
         return web.json_response()
 
+    @instance.routes.get("/sd-webui-comfyui/default_workflow")
+    async def get_default_workflow(request):
```
Comment (ljleb) [inline lib_comfyui/polling_client.py:139]: this function should be defined in its own file, most likely. the hijack logic in this file could be moved somewhere else?

File: javascript/compute-height.js [hunk 7]
```diff
@@ -5,6 +5,16 @@ function uuidv4() {
 }
 
 const CLIENT_KEY = uuidv4();
+
+function changeCurrentWorkflow(workflowTypes, newWorkflowName) {
```
Comment (John-WL) [inline javascript/compute-height.js:9]: can we rename this to `changeDisplayedWorkflow`?

File: comfyui_custom_scripts/extensions/webuiRequests.js [hunk 8]
```diff
@@ -73,14 +73,15 @@ function onElementDomIdRegistered(callback) {
 }
 
 function patchUiEnv(thisClientId) {
-    if(thisClientId !== 'comfyui_general_tab') {
+    if(thisClientId !== 'sandbox_tab') {
```
Comment (John-WL) [inline comfyui_custom_scripts/extensions/webuiRequests.js:76]: I know all the ids have comfyui in their names, but doing this confuses me more than anything. In my view, the variable should contain the full name in order to make things clear that it represents directly the HTML id property. What do you think?

File: javascript/compute-height.js [hunk 9]
```diff
@@ -129,22 +128,21 @@ function reloadFrameElement(iframeElement) {
     iframeElement.src = oldSrc;
 }
 
-function forceFeedIdToIFrame(frameId) {
+function forceFeedIdToIFrame(workflowId) {
```
Comment (John-WL) [inline javascript/compute-height.js:129]: I don't think we can rename this to `workflowId` because it's not the workflow that has an id (a "workflow" is a graph of nodes), it's the iframe itself. In my view, the iframeId targets a specific iframe

File: lib_comfyui/external_code.py [hunk 10]
```diff
@@ -91,12 +91,12 @@ def get_workflow_display_names(tabs: Tabs = ALL_TABS) -> List[str]:
     return [workflow.display_name for workflow in get_workflows(tabs)]
 
 
-def get_default_workflow_json(iframe_id: str) -> dict:
+def get_default_workflow_json(workflow_id: str) -> dict:
```
Comment (John-WL) [inline lib_comfyui/external_code.py:96]: in my view, again, this should stay named as `iframe_id` (or renamed as `workflow_type_id`)

File: javascript/compute-height.js (hunk 9 above)
Comment (John-WL) [reply javascript/compute-height.js:129]: My understanding is that there is a difference between workflows and worflow types, and it's important to see it in the code.

Comment (ljleb) [reply javascript/compute-height.js:129]: Sure, I can rename the concept of workflows to workflow types as it overlaps with the existing concept of comfyui workflow. Would that make more sense in your view?

File: javascript/compute-height.js [hunk 11]
```diff
@@ -114,8 +112,8 @@ function getComfyuiContainer() {
     return document.getElementById("comfyui_webui_container") ?? null;
 }
 
-function getComfyuiIFrameElement() {
-    return document.querySelector('comfyui_general_tab') ?? null;
+function getWorkflowIdsElement() {
+    return document.getElementById('comfyui_workflow_id_list') ?? null;
```
Comment (John-WL) [inline javascript/compute-height.js:116]: same again, I would call this `getWorkflowTypeIdsElement` or leave it as `getIFrameIdsElement`

File: lib_comfyui/webui_callbacks.py [hunk 12]
```diff
@@ -1,5 +1,6 @@
+from lib_comfyui import comfyui_adapter, webui_tab, webui_settings, ipc, comfyui_context
+ipc.restrict_to_process('webui')(lambda: None)()
```
Comment (John-WL) [inline lib_comfyui/webui_callbacks.py]: what is the point of this lol

File: lib_comfyui/external_code.py [hunk 13]
```diff
@@ -37,69 +37,80 @@ def get_ids(self, tabs: Tabs = ALL_TABS) -> List[str]:
...
+    set_workflow_types(workflows)
 
 
-def get_workflows(tabs: Tabs = ALL_TABS) -> List[Workflow]:
+def get_workflow_types(tabs: Tabs = ALL_TABS) -> List[WorkflowType]:
     """
     Get the list of currently registered workflows
-    To update the workflows, do not modify this list directly
-    Instead, have a look at the other functions
+    To update the workflows list, see `add_workflow` or `set_workflows`
```
Comment (John-WL) [inline lib_comfyui/external_code.py:63]: PladsParseError: comment points to a non-existent function :p

File: javascript/compute-height.js [hunk 14]
```diff
@@ -29,31 +32,35 @@ function onComfyuiTabLoaded(callback) {
     callback();
 }
 
+let WORKFLOW_TYPE_IDS = [];
+
 function setupComfyuiTabEvents() {
+    WORKFLOW_TYPE_IDS = JSON.parse(getWorkflowTypeIdsElement().innerText);
```
Comment (ljleb) [inline javascript/compute-height.js:38]: consider making a fastapi route instead of this hack

File: javascript/compute-height.js [hunk 15]
```diff
@@ -119,22 +130,21 @@ function reloadFrameElement(iframeElement) {
...
-
     });
 
     const feed = () => {
         if(received) return;
-        const frameEl = document.querySelector(`#${frameId}`);
+        const frameEl = getWorkflowTypeIFrame(workflowId);
         const targetOrigin = frameEl.src;
-        const message = `${frameEl.getAttribute('id')}.${CLIENT_KEY}`;
+        const message = `${workflowId}.${WEBUI_CLIENT_KEY}`;
```
Comment (ljleb) [inline javascript/compute-height.js:147]: workflowTypeId

File: lib_comfyui/comfyui/pre_main.py [hunk 16]
```diff
@@ -3,20 +3,21 @@
...
+from lib_comfyui.comfyui import routes_extension, queue_tracker
+from lib_comfyui.webui import paths
+
+
+original_print = builtins.print
+def comfyui_print(*args, **kwargs):
+    return original_print('[ComfyUI]', *args, **kwargs)
 
 
 def main(comfyui_path, webui_folder_paths, process_queues, cli_args):
```
Comment (ljleb) [inline lib_comfyui/comfyui/pre_main.py:19]: webui_folder_paths is not necessary here now that we have ipc

**Comment by ljleb** [comment]
This is starting to be pretty big. Maybe we should split it in smaller pieces.

Tested basic workflows in WSL and windows 10.

**Comment by ljleb** [comment]
I think the readme and wiki should be updated to show examples of `external_code` module usages.

File: comfyui_custom_nodes/webui_checkpoint_loader.py [hunk 17]
```diff
@@ -23,12 +15,12 @@ def INPUT_TYPES(s):
...
-        webui_proxies.raise_on_unsupported_model_type(config)
+        config = proxies.get_comfy_model_config()
+        proxies.raise_on_unsupported_model_type(config)
         return (
-            WebuiModelPatcher(WebuiModelProxy()),
-            WebuiClipWrapper(WebuiClipProxy()),
-            WebuiVaeWrapper(WebuiVaeProxy()),
+            proxies.ModelPatcher(proxies.Model()),
+            proxies.ClipWrapper(proxies.Clip()),
+            proxies.VaeWrapper(proxies.Vae()),
```
Comment (John-WL) [inline comfyui_custom_nodes/webui_checkpoint_loader.py:23]: do you think it would be a better idea to pass these values by default?

Comment (ljleb) [reply comfyui_custom_nodes/webui_checkpoint_loader.py:23]: I did consider it, but I'm not sure whether by default or not is better. Requiring to pass these values means exposing the dependencies, which should ensure ease of testing if we were to ever add unit tests for these classes.

File: comfyui_custom_scripts/extensions/webuiRequests.js [hunk 18]
```diff
@@ -110,11 +99,57 @@ function patchSavingMechanism() {
...
+            return app.original_loadGraphData(defaultGraph);
+        } else {
+            return app.original_loadGraphData(graphData);
+        }
+    };
+
+    app.loadGraphData();
+}
+
+export const iframeRegisteredEvent = new Promise(resolve => {
```
Comment (John-WL) [inline comfyui_custom_scripts/extensions/webuiRequests.js:131]: do you think it would be a better idea to specify that this is an exported module component at the end of the file? 

```js
export {
    iframeRegisteredEvent,
}
```

Comment (John-WL) [reply comfyui_custom_scripts/extensions/webuiRequests.js:131]: like that, it becomes easier to track what is and isn't exported, and it's easier to modify in one place

Comment (ljleb) [reply comfyui_custom_scripts/extensions/webuiRequests.js:131]: We could also move it to a different file and import it instead maybe? This one and `appGraphEvent` below.

File: comfyui_custom_scripts/extensions/webuiNodes.js [hunk 19]
```diff
@@ -30,6 +26,16 @@ const ext = {
...
         };
     },
+    async beforeRegisterNodeDef(node, nodeData) {
+        const iframeInfo = await iframeRegisteredEvent;
+
+        if (!Object.keys(iframeInfo.webuiIoNodeDisplayNameMappings).includes(nodeData.name)) {
+            return;
+        }
+
+        nodeData.display_name = `${nodeData.display_name} - ${iframeInfo.workflowTypeDisplayName}`;
```
Comment (John-WL) [inline comfyui_custom_scripts/extensions/webuiNodes.js:36]: why is there a `iframeInfo.webuiIoNodeDisplayNameMappings` attribute if we don't use the map and directly use `iframeInfo.workflowTypeDisplayName` instead? 

I get that `webuiIoNodeDisplayNameMappings` is usefull for filtering the nodes. My understanding is that this is a python dict that's translated into a js object. If I change the python dict, then these changes will be reflected in the object, but we are still going to use `workflowTypeDisplayName`, and so the changes won't have any effect. 

This can create problems because it *appears* like the dict in python is a map, but we only use the keys. 

All of this is assuming that the value of the dict is stored somewhere else. This is assumed because the expected behavior here would be to use `iframeInfo.webuiIoNodeDisplayNameMappings[nodeData.name]`, but it wasn't used, and so this means that the dict values are not mapped and are distinct from the possible `iframeInfo.workflowTypeDisplayName`. 

Sorry if this is long, I wanted to make sure to make my point clear! Let me know if I'm misunderstanding anything, or if we should send the values separately from the dict even though we can directly map them with the dict itself.

Comment (John-WL) [reply comfyui_custom_scripts/extensions/webuiNodes.js:36]: Mhh let me rephrase. My problem is that there are 2 possible ways to get the right name for the node:
1)
```js
        nodeData.display_name = `${nodeData.display_name} - ${iframeInfo.workflowTypeDisplayName}`;
```
2)
```js
        nodeData.display_name = `${nodeData.display_name} - ${iframeInfo.webuiIoNodeDisplayNameMappings[nodeData.name]}`;
```

why not use the 2nd one, and NOT pass the value? If we know the value, why not compare it directly with `nodeData.name` instead of cheking if it's in the map?

File: comfyui_custom_nodes/webui_checkpoint_loader.py (hunk 17 above)
Comment (ljleb) [reply comfyui_custom_nodes/webui_checkpoint_loader.py:23]: Also, `proxies.ModelPatcher` must receive the model as argument, since `proxies.ClipWrapper` also uses it.

File: lib_comfyui/comfyui/routes_extension.py [hunk 20]
```diff
@@ -0,0 +1,150 @@
+import asyncio
+import json
+from typing import List
+import torch
+from lib_comfyui import parallel_utils, ipc, global_state, comfyui_context, torch_utils, external_code
+from lib_comfyui.comfyui import queue_tracker
+
+
+# rest is ran on comfyui's server side
+class ComfyuiNodeWidgetRequests:
```
Comment (John-WL) [inline lib_comfyui/comfyui/routes_extension.py:10]: We could probably rename this to `ComfyuiIFrameRequests`

File: lib_comfyui/webui/paths.py [hunk 21]
```diff
@@ -1,19 +1,21 @@
...
     for folder_id, folder_paths in folder_paths.items():
         for folder_path in folder_paths:
             add_model_folder_path(folder_id, folder_path)
 
 
-def get_folder_paths() -> dict:
+@ipc.run_in_process('webui')
+def get_webui_folder_paths() -> dict:
+    from modules import paths, shared
     from modules import sd_models
```
Comment (John-WL) [inline lib_comfyui/webui/paths.py:19]:
```py
    from modules import paths, shared, sd_models
```
?

File: lib_comfyui/webui/workflow_patcher.py [hunk 22]
```diff
@@ -0,0 +1,85 @@
...
+import torch
+import torchvision.transforms.functional as F
+from lib_comfyui import ipc, global_state, default_workflow_types, external_code
+
+
+__original_create_sampler = None
+
+
+@ipc.restrict_to_process('webui')
+def apply_patches():
```
Comment (John-WL) [inline lib_comfyui/webui/workflow_patcher.py:11]:
```py
def apply_patch():
```
?

I mean, I only see one patch, and I would expect a for loop with that name

Comment (ljleb) [reply lib_comfyui/webui/workflow_patcher.py:11]: If we want to add more in the future, this is where we would put them.

File: lib_comfyui/private/external_code.py [hunk 23]
```diff
@@ -0,0 +1,203 @@
...
+        if isinstance(self.tabs, str):
+            self.tabs = (self.tabs,)
+
+        assert self.tabs, "tabs must not be empty"
+
+        if isinstance(self.default_workflow, Path):
+            with open(str(self.default_workflow), 'r') as f:
+                self.default_workflow = f.read()
+
+    def get_ids(self, tabs: Tabs = ALL_TABS) -> List[str]:
```
Comment (John-WL) [inline lib_comfyui/private/external_code.py:32]: can you explain why each and every workflow type can have many ids?

Comment (ljleb) [reply lib_comfyui/private/external_code.py:32]: One id per tab. For example, the Postprocess workflow type has ids `postprocess_txt2img` and `postprocess_img2img`

Comment (ljleb) [reply lib_comfyui/private/external_code.py:32]: Both are the same workflow type, but they appear on different tabs and so they need different ids so that python can target a single one of them at a time.

File: scripts/comfyui.py [hunk 24]
```diff
@@ -1,71 +1,122 @@
...
+    def get_iframes_html(self, is_img2img: bool) -> str:
+        comfyui_client_url = settings.get_comfyui_client_url()
+
+        iframes = []
+        first_loop = True
+        for workflow_type_id in external_code.get_workflow_type_ids(self.get_xxx2img_str(is_img2img)):
+            html_classes = ['comfyui-embedded-widget']
+            if first_loop:
+                first_loop = False
+                html_classes.append('comfyui-embedded-widget-display')
```
Comment (John-WL) [inline scripts/comfyui.py:81]: let me know if I'm wrong, but I think it would be better to use the default element that was set in the `gr.Dropdown():` as the first displayed value here

File: scripts/comfyui.py [hunk 25]
```diff
@@ -1,71 +1,122 @@
...
-            batch=pp.images,
-            workflow_type='postprocess',
-            is_img2img=self.is_img2img,
-            required_node_types=webui_postprocess_output.expected_node_types,
-            queue_front=queue_front,
+        batch_results = external_code.run_workflow(
+            workflow_type=default_workflow_types.postprocess_workflow_type,
+            tab=self.get_xxx2img_str(),
+            batch_input=pp.images,
         )
```
Comment (John-WL) [inline scripts/comfyui.py:117]: do you think it would be a good idea to use our own API instead of bypassing it to run workflow types? I'm talking about the code in `external_code.py`

File: scripts/comfyui.py (hunk 24 above)
Comment (ljleb) [reply scripts/comfyui.py:81]: Sure, I can refact the condition. It's the first one in the dropdown so it has the same behavior.

**Review by John-WL — APPROVED** [review]
